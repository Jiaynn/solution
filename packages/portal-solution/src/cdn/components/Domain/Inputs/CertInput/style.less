@import "../common/style.less";

.domain-cert-input-wrapper {
  .input-wrapper();

  .tip-icon {
    margin-left: 1em;
  }

  .cert-input-tips {
    display: block;
    margin-left: 0;
    margin-top: 8px;
    font-size: 14px;
    color: #999;
  }

  .cert-detail {
    font-size: 14px;

    .line.cert-detail-item {
      color: #999;
    }
  }
}
