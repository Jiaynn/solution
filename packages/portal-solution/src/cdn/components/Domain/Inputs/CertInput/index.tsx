/**
 * @file 指定 domain 的 cert 选择
 * @author nighca <nighca@live.cn>
 */

import React from 'react'
import moment from 'moment'
import { uniqBy } from 'lodash'
import { observable, action, reaction, computed } from 'mobx'
import { observer } from 'mobx-react'
import autobind from 'autobind-decorator'
import Select from 'react-icecream/lib/select'
import { injectProps } from 'qn-fe-core/local-store'
import Store, { observeInjectable as injectable } from 'qn-fe-core/store'
import { Link } from 'portal-base/common/router'
import { IamDisabled } from 'portal-base/user/iam'
import { DomainApis, ICertSearchByDomain } from 'portal-base/fusion'
import { SslClient } from 'portal-base/certificate'
import { useLocalStore } from 'portal-base/common/utils/store'

import { humanizeTimeUTC, getDiffDays } from 'cdn/transforms/datetime'
import { convertCertInfoToCertSearchInfo } from 'cdn/transforms/domain'

import { domainWithWildcard as domainPattern } from 'cdn/constants/pattern'

import OEMDisabled from 'cdn/components/common/OEMDisabled'

import Error from '../common/Error'

import './style.less'

export interface IDomainCertInputProps {
  domain: string
  value: string
  error?: any
  domainCertId?: string
  uid?: number
  onChange: (value: string) => void
}

@injectable()
export class LocalStore extends Store {

  @observable.ref certs: ICertSearchByDomain[] = []
  @observable.ref domainCert?: ICertSearchByDomain

  @action updateCerts(certs: ICertSearchByDomain[]) {
    this.certs = certs || []
  }

  @computed get allCerts() {
    // 证书列表的接口不会返回有效期小于 7 天的证书
    const certs = this.domainCert ? this.certs.concat(this.domainCert) : this.certs
    return uniqBy(certs, it => it.id)
  }

  fetchCerts(domain: string, uid?: number) {
    return this.domainApis.searchCertsByDomain({ domain, uid }).then(
      certs => this.updateCerts(certs)
    )
  }

  @action.bound updateDomainCertInfo(info?: ICertSearchByDomain) {
    this.domainCert = info
  }

  fetchDomainCertInfo(certId: string) {
    return this.sslClient.getCertInfo(certId).then(
      resp => this.updateDomainCertInfo(convertCertInfoToCertSearchInfo(resp))
    )
  }

  @computed get selectedCert() {
    const selectedCertId = this.props.value
    if (!selectedCertId) {
      return null
    }
    return this.allCerts.find(
      cert => cert.id === selectedCertId
    )
  }

  constructor(
    @injectProps() private props: IDomainCertInputProps,
    private sslClient: SslClient,
    private domainApis: DomainApis
  ) {
    super()
  }

  init() {
    // domain 内容有变，则重新加载对应的可选 cert 列表
    this.addDisposer(reaction(
      () => this.props.domain,
      domain => {
        this.updateCerts([])
        if (domain && domainPattern.test(domain)) {
          this.fetchCerts(domain)
        }
      },
      { fireImmediately: true }
    ))

    this.addDisposer(reaction(
      () => this.props.domainCertId,
      certId => {
        this.updateDomainCertInfo()
        if (certId) {
          this.fetchDomainCertInfo(certId)
        }
      },
      { fireImmediately: true }
    ))

    // certs 列表有变化且当前选中项不合法时，自动更新值
    this.addDisposer(reaction(
      () => this.allCerts,
      certs => {
        if (!this.selectedCert) {
          this.props.onChange(
            certs.length > 0
            ? certs[0].id
            : ''
          )
        }
      }
    ))
  }
}

type PropsWithDeps = IDomainCertInputProps & {
  store: LocalStore
}

@observer
class DomainCertInputInner extends React.Component<PropsWithDeps> {

  @autobind handleCertChange(certId: string) {
    this.props.onChange(certId)
  }

  getCertDetail() {
    const selectedCert = this.props.store.selectedCert
    if (!selectedCert) {
      return null
    }

    const dnsnameItems = selectedCert.dnsnames || []

    const [notBefore, notAfter] = [selectedCert.not_before, selectedCert.not_after].map(
      timeText => humanizeTimeUTC(timeText)
    )

    return (
      <div className="cert-detail">
        <div className="line cert-detail-item">
          证书包含域名：{dnsnameItems.join(' | ')}
        </div>
        <div className="line cert-detail-item">
          证书有效期：{notBefore} ~ {notAfter}
        </div>
      </div>
    )
  }

  getCertInputTips() {
    return (
      <div className="cert-input-tips">
        <IamDisabled>
          <OEMDisabled>
            您可以在&nbsp;<Link to="/certificate/ssl#cert">SSL 证书服务</Link>&nbsp;页面申请或上传自有证书。
          </OEMDisabled>
        </IamDisabled>
        该列表不包含有效期小于 7 天的证书。
      </div>
    )
  }

  render() {
    const certs = this.props.store.allCerts
    const options = certs.map(
      cert => (
        <Select.Option disabled={IsInvalidCert(cert)} key={cert.id} value={cert.id}>
          {humanizeCertInfo(this.props.domainCertId!, cert)}
        </Select.Option>
      )
    )

    // null/undefined 在这里都是无效的值，统一处理成 undefined 是为了显示 placeholder
    const value = this.props.value != null
      ? this.props.value
      : undefined

    return (
      <div className="domain-cert-input-wrapper">
        <div className="line">
          <div className="text-input-wrapper">
            <Select
              disabled={certs.length === 0}
              value={value}
              onChange={this.handleCertChange}
              placeholder={certs.length > 0 ? '请选择证书' : '暂无证书'}
              dropdownStyle={{ minWidth: '360px' }}
            >{options}</Select>
          </div>
          <Error error={this.props.error} />
          {this.getCertInputTips()}
        </div>
        {this.getCertDetail()}
      </div>
    )
  }
}

function getCertExpireDays(cert: ICertSearchByDomain) {
  return getDiffDays(moment(), moment(cert.not_after))
}

// 有效期小于 7 天被认为是无效的证书，即：不能被选中
function IsInvalidCert(cert: ICertSearchByDomain) {
  const diffDays = getCertExpireDays(cert)
  return diffDays < 7
}

function getCertExpireInfo(cert: ICertSearchByDomain) {
  const diffDays = getCertExpireDays(cert)
  if (diffDays < 0) {
    return '证书已过期'
  }
  return `有效期 ${diffDays} 天`
}

function humanizeCertInfo(currentCertId: string, cert: ICertSearchByDomain) {
  const certName = cert.cert_name || cert.name
  const expireInfo = getCertExpireInfo(cert)
  const currentCertInfo = currentCertId === cert.id ? '（当前绑定）' : ''

  return [certName, expireInfo, currentCertInfo].filter(Boolean).join(' ')
}

export default function DomainCertInput(props: IDomainCertInputProps) {
  const store = useLocalStore(LocalStore, props)
  return (
    <DomainCertInputInner {...props} store={store} />
  )
}
