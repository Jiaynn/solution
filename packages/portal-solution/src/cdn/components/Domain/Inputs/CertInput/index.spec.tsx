import React from 'react'
import { PROPS } from 'qn-fe-core/local-store'

import { createContainer, RendererUtils as Renderer } from 'test'
import CertInput, { LocalStore } from '.'

const storeProps = {
  domain: 'qiniu.com',
  value: 'true',
  onChange: jest.fn()
}

const storePropsWithError = {
  domain: 'qiniu.com',
  value: 'true',
  error: 'mortal error',
  onChange: jest.fn()
}

const cert = {
  id: '3213',
  not_before: '20111030',
  not_after: '20180330',
  dnsnames: ['qiniu.com', 'qiniu.io'],
  name: 'qiniu',
  cert_name: 'qiniu biz'
}

it('store update correctly', () => {
  const container = createContainer([
    LocalStore,
    { identifier: PROPS, value: storeProps }
  ])
  const store = container.get(LocalStore)
  const certs = [cert]
  store.updateCerts(certs)
  expect(store.certs).toEqual(certs)
})

const renderer = new Renderer()

it('renders correctly', () => {
  const tree = renderer.createWithAct(
    <CertInput {...storeProps} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly with error', () => {
  const tree = renderer.createWithAct(
    <CertInput {...storePropsWithError} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
