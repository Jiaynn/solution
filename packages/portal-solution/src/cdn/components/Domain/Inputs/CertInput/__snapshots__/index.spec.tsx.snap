// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="domain-cert-input-wrapper"
>
  <div
    className="line"
  >
    <div
      className="text-input-wrapper"
    >
      <div
        className="ant-select ant-select-disabled"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        onMouseOut={[Function]}
        onMouseUp={[Function]}
      >
        <div
          aria-autocomplete="list"
          aria-controls="test-uuid"
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-select-selection
            ant-select-selection--single"
          onKeyDown={[Function]}
          role="combobox"
          tabIndex={-1}
        >
          <div
            className="ant-select-selection__rendered"
          >
            <div
              className="ant-select-selection__placeholder"
              onClick={[Function]}
              onMouseDown={[Function]}
              style={
                Object {
                  "WebkitUserSelect": "none",
                  "display": "none",
                  "userSelect": "none",
                }
              }
              unselectable="on"
            >
              暂无证书
            </div>
            <div
              className="ant-select-selection-selected-value"
              style={
                Object {
                  "display": "block",
                  "opacity": 1,
                }
              }
              title="true"
            >
              true
            </div>
          </div>
          <span
            className="ant-select-arrow"
            onClick={[Function]}
            style={
              Object {
                "WebkitUserSelect": "none",
                "userSelect": "none",
              }
            }
            unselectable="on"
          >
            <i
              aria-label="icon: down"
              className="anticon anticon-down ant-select-arrow-icon"
            >
              <svg
                aria-hidden="true"
                className=""
                data-icon="down"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                />
              </svg>
            </i>
          </span>
        </div>
      </div>
    </div>
    <div
      className="cert-input-tips"
    >
      您可以在 
      <a
        className="comp-router-link"
        href="/certificate/ssl#cert"
        onClick={[Function]}
      >
        SSL 证书服务
      </a>
       页面申请或上传自有证书。
      该列表不包含有效期小于 7 天的证书。
    </div>
  </div>
</div>
`;

exports[`renders correctly with error 1`] = `
<div
  className="domain-cert-input-wrapper"
>
  <div
    className="line"
  >
    <div
      className="text-input-wrapper"
    >
      <div
        className="ant-select ant-select-disabled"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        onMouseOut={[Function]}
        onMouseUp={[Function]}
      >
        <div
          aria-autocomplete="list"
          aria-controls="test-uuid"
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-select-selection
            ant-select-selection--single"
          onKeyDown={[Function]}
          role="combobox"
          tabIndex={-1}
        >
          <div
            className="ant-select-selection__rendered"
          >
            <div
              className="ant-select-selection__placeholder"
              onClick={[Function]}
              onMouseDown={[Function]}
              style={
                Object {
                  "WebkitUserSelect": "none",
                  "display": "none",
                  "userSelect": "none",
                }
              }
              unselectable="on"
            >
              暂无证书
            </div>
            <div
              className="ant-select-selection-selected-value"
              style={
                Object {
                  "display": "block",
                  "opacity": 1,
                }
              }
              title="true"
            >
              true
            </div>
          </div>
          <span
            className="ant-select-arrow"
            onClick={[Function]}
            style={
              Object {
                "WebkitUserSelect": "none",
                "userSelect": "none",
              }
            }
            unselectable="on"
          >
            <i
              aria-label="icon: down"
              className="anticon anticon-down ant-select-arrow-icon"
            >
              <svg
                aria-hidden="true"
                className=""
                data-icon="down"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                />
              </svg>
            </i>
          </span>
        </div>
      </div>
    </div>
    <span
      className="error-wrapper"
      style={
        Object {
          "color": "#e55c5c",
          "display": "block",
          "lineHeight": "20px",
          "marginTop": "8px",
          "verticalAlign": "middle",
        }
      }
    >
      mortal error
    </span>
    <div
      className="cert-input-tips"
    >
      您可以在 
      <a
        className="comp-router-link"
        href="/certificate/ssl#cert"
        onClick={[Function]}
      >
        SSL 证书服务
      </a>
       页面申请或上传自有证书。
      该列表不包含有效期小于 7 天的证书。
    </div>
  </div>
</div>
`;
